### DISCUSSION TITLE
SelectionPrompt/AddChoice use a Tree (put only the text)

### DISCUSSION THREAD

**Discussion by DiagProf** [spectreconsole/spectre.console#550] [Q&A] [answered]
I would like to use the tree (text would be enough) to enrich the selection prompt with more information. 
Can that be achieved somehow?

That's roughly how I would imagine the representation.

![image](https://user-images.githubusercontent.com/56319388/134785009-e1359ad1-44e1-4898-95b4-556a2c61a0fa.png)

this is my source code i'm playing with but it doesn't work.

using System;
using Spectre.Console;

namespace SpectreConsoleTest
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("Welcome to the rental park of fun vehicles.");
            Console.WriteLine("We have everything you always wanted to drive.");
            Console.WriteLine();

            var rootCaterpillar = new Tree("Caterpillar D8T");
            // Add some nodes
            var fooCaterpillar = rootCaterpillar.AddNode("[yellow]Engine[/]");
            fooCaterpillar.AddNode("Engine Model Cat C15");
            fooCaterpillar.AddNode("354 HP");
            
            var barCaterpillar = rootCaterpillar.AddNode("[yellow]Body shape[/]");
            barCaterpillar.AddNode("Operating Weight 87600 lb");
            barCaterpillar.AddNode("Yellow");

            var rootKraussMaffei = new Tree("Krauss-Maffei Leopard 2");
            // Add some nodes
            var fooKraussMaffei = rootKraussMaffei.AddNode("[yellow]Engine[/]");
            fooKraussMaffei.AddNode("12-Zylinder-Dieselengine MTU");
            fooKraussMaffei.AddNode("CO2 emissions don't care");

            var barKraussMaffei = rootKraussMaffei.AddNode("[yellow]Military equipment[/]");
            barKraussMaffei.AddNode("Laser sword");
            barKraussMaffei.AddNode("Full rotating tower");

            var selectionPrompt = new SelectionPrompt<Tree>()
                    .Title("What's your [blue]favorite fun vehicle[/]?")
                    .PageSize(10)
                    .AddChoices(rootCaterpillar, rootKraussMaffei);

            selectionPrompt.Converter = tree =>
            {
                //I'm stuck here
                return tree.ToString();
            };

            var rentalVehicle = AnsiConsole.Prompt<Tree>(
                selectionPrompt
            );

        }
    }
}

**Comment by nils-a** [comment]
Is there a particular reason why you're trying to convert a tree to a prompt?

When you add child-nodes to a `SelectionPrompt`, I think the output should resemble what you are trying to achieve.

For example, the code

```cs
using System;
using Spectre.Console;

namespace SpectreConsoleTest
{
    public static class Program
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("Welcome to the rental park of fun vehicles.");
            Console.WriteLine("We have everything you always wanted to drive.");
            Console.WriteLine();

            var prompt = new SelectionPrompt<string>
            {
                PageSize = 10,
                Title = "What's your [blue]favorite fun vehicle[/]?",
                Mode = SelectionMode.Leaf
            };
            
            var rootCaterpillar = prompt.AddChoice("Caterpillar D8T");
            // Add some nodes
            var fooCaterpillar = rootCaterpillar.AddChild("[yellow]Engine[/]");
            fooCaterpillar.AddChild("Engine Model Cat C15");
            fooCaterpillar.AddChild("354 HP");

            var barCaterpillar = rootCaterpillar.AddChild("[yellow]Body shape[/]");
            barCaterpillar.AddChild("Operating Weight 87600 lb");
            barCaterpillar.AddChild("Yellow");

            var rootKraussMaffei = prompt.AddChoice("Krauss-Maffei Leopard 2");
            // Add some nodes
            var fooKraussMaffei = rootKraussMaffei.AddChild("[yellow]Engine[/]");
            fooKraussMaffei.AddChild("12-Zylinder-Dieselengine MTU");
            fooKraussMaffei.AddChild("CO2 emissions don't care");

            var barKraussMaffei = rootKraussMaffei.AddChild("[yellow]Military equipment[/]");
            barKraussMaffei.AddChild("Laser sword");
            barKraussMaffei.AddChild("Full rotating tower");

            var rentalVehicle = AnsiConsole.Prompt(prompt);
            AnsiConsole.MarkupLine($"You selected [green]{rentalVehicle}[/]");

        }
    }
}
```

produces a prompt like:

![image](https://user-images.githubusercontent.com/349188/134785228-76e6c24b-e375-48a6-a176-ce470bf13df3.png)

**Comment by DiagProf** [comment, marked as answer]
@nils-a  Thank you for your prompt reply.  

In my invented example, only the "Caterpillar D8T" and "Krauss-Maffei Leopard 2" vehicles can be rented. For the user (especially the newcomers) I would like to give a few more meta data (in the tree view) to get a feel for what kind of vehicle it is.

Could also think of network printers that are standing around somewhere in the company. And the metadata says where it is located in the company. After printing twice, the user no longer reads this and the root name is sufficient, e.g. "Brother MFC-7820N USB Printer"

but your idea would work if there was something like that

                var prompt = new SelectionPrompt<string>
                {
                    PageSize = 10,
                    Title = "What's your [blue]favorite fun vehicle[/]?",
                    Mode = SelectionMode.**RootOnly**  
                };
but there is currently no **RootOnly**

**Reply by nils-a** [reply]
Interesting. I could have seen that, if I had read your post more carefully. Sorry for that.

I don't think that's easily possible. While `Prompt` can use arbitrary structures, the `Convert` method needs to [return a `string`](https://github.com/spectreconsole/spectre.console/blob/fa5a1e88ec4211d69a24df491999fa02e3e80188/src/Spectre.Console/Prompts/SelectionPrompt.cs#L44). 

As an idea to get you started, you could use something along the lines of 
```cs
using System;
using System.Linq;
using System.Text;
using Spectre.Console;
using Spectre.Console.Rendering;

namespace SpectreConsoleTest
{
    public static class Program
    {

        class SimpleCapabilities : IReadOnlyCapabilities
        {
            // todo: read somehow from console?
            public ColorSystem ColorSystem { get; } = ColorSystem.Standard;
            public bool Ansi { get; } = true;
            public bool Links { get; } = true;
            public bool Legacy { get; } = false;
            public bool IsTerminal { get; } = true;
            public bool Interactive { get; } = false;
            public bool Unicode { get; } = true;
        }
        
        public static void Main(string[] args)
        {
            Console.WriteLine("Welcome to the rental park of fun vehicles.");
            Console.WriteLine("We have everything you always wanted to drive.");
            Console.WriteLine();

            var simpleRenderContext = new RenderContext(new SimpleCapabilities(), Justify.Left);
            var width = 80; // todo: read somehow from console

            var prompt = new SelectionPrompt<Tree>
            {
                Converter = v =>
                {
                    var sb = new StringBuilder();
                    var segments = ((IRenderable)v).Render(simpleRenderContext, width);
                    foreach (var segment in segments)
                    {
                        sb.Append(segment.Text);
                    }

                    return sb.ToString();
                },
                PageSize = 10,
                Title = "What's your [blue]favorite fun vehicle[/]?",
                Mode = SelectionMode.Leaf
            };
            
            var rootCaterpillar = new Tree("Caterpillar D8T");
            // Add some nodes
            var fooCaterpillar = rootCaterpillar.AddNode("[yellow]Engine[/]");
            fooCaterpillar.AddNode("Engine Model Cat C15");
            fooCaterpillar.AddNode("354 HP");

            var barCaterpillar = rootCaterpillar.AddNode("[yellow]Body shape[/]");
            barCaterpillar.AddNode("Operating Weight 87600 lb");
            barCaterpillar.AddNode("Yellow");
            prompt.AddChoice(rootCaterpillar);

            var rootKraussMaffei =  new Tree("Krauss-Maffei Leopard 2");
            // Add some nodes
            var fooKraussMaffei = rootKraussMaffei.AddNode("[yellow]Engine[/]");
            fooKraussMaffei.AddNode("12-Zylinder-Dieselengine MTU");
            fooKraussMaffei.AddNode("CO2 emissions don't care");

            var barKraussMaffei = rootKraussMaffei.AddNode("[yellow]Military equipment[/]");
            barKraussMaffei.AddNode("Laser sword");
            barKraussMaffei.AddNode("Full rotating tower");
            prompt.AddChoice(rootKraussMaffei);
            
            var rentalVehicle = AnsiConsole.Prompt(prompt);
            AnsiConsole.MarkupLine($"You selected [green]{rentalVehicle}[/]");

        }
    }
}
```

However, all the coloring is lost this way. (And I'm not really sure if this is an intended use.)

**Reply by DiagProf** [reply]
many thanks. That's basically what i was looking for.

I still manipulated the formatting here. 

                    foreach (var segment in segments)
                    {
                        sb.Append(segment.Text);
                        
                        //i add this lines
                        if (segment.IsLineBreak)
                            sb.Append("  "); 
                    }

And now it looks like I imagined.
![image](https://user-images.githubusercontent.com/56319388/134801948-322d235b-4167-4eec-a534-0d09a786bdcd.png)


Now there is only one last problem, how do I get the label back from the root node in this line?
AnsiConsole.MarkupLine($"You selected [green]{rentalVehicle}[/]");
I need something like rentalVehicle.root label

**Reply by nils-a** [reply]
Actually, I would add a class that encapsulates your choices (and probably their properties) and use the tree only to get the correct formatting. 

Something like:

```cs
using System;
using System.Linq;
using System.Text;
using Spectre.Console;
using Spectre.Console.Rendering;

namespace SpectreConsoleTest
{
    public static class Program
    {

        class SimpleCapabilities : IReadOnlyCapabilities
        {
            // todo: read somehow from console?
            public ColorSystem ColorSystem { get; } = ColorSystem.Standard;
            public bool Ansi { get; } = true;
            public bool Links { get; } = true;
            public bool Legacy { get; } = false;
            public bool IsTerminal { get; } = true;
            public bool Interactive { get; } = false;
            public bool Unicode { get; } = true;
        }
        
        public class Vehicle
        {
            public Vehicle(string title, Action<Tree> additionalInformation)
            {
                Title = title;
                AdditionalInformation = additionalInformation;
            }
            
            public string Title { get; set; }
            public Action<Tree> AdditionalInformation { get; set; }

            public string GetFormattedPrompt()
            {
                var tree = new Tree(Title);
                AdditionalInformation(tree);
                
                var simpleRenderContext = new RenderContext(new SimpleCapabilities(), Justify.Left);
                var width = 80; // todo: read somehow from console
                var sb = new StringBuilder();
                var segments = ((IRenderable)tree).Render(simpleRenderContext, width);
                foreach (var segment in segments)
                {
                    sb.Append(segment.Text);

                    if (segment.IsLineBreak)
                        sb.Append("  ");
                }

                return sb.ToString();
            }
        }
        
        public static void Main(string[] args)
        {
            Console.WriteLine("Welcome to the rental park of fun vehicles.");
            Console.WriteLine("We have everything you always wanted to drive.");
            Console.WriteLine();

            var prompt = new SelectionPrompt<Vehicle>
            {
                Converter = v => v.GetFormattedPrompt(),
                PageSize = 10,
                Title = "What's your [blue]favorite fun vehicle[/]?",
                Mode = SelectionMode.Leaf
            };
            
            prompt.AddChoice(new Vehicle("Caterpillar D8T", rootCaterpillar =>
            {
                // Add some nodes
                var fooCaterpillar = rootCaterpillar.AddNode("[yellow]Engine[/]");
                fooCaterpillar.AddNode("Engine Model Cat C15");
                fooCaterpillar.AddNode("354 HP");

                var barCaterpillar = rootCaterpillar.AddNode("[yellow]Body shape[/]");
                barCaterpillar.AddNode("Operating Weight 87600 lb");
                barCaterpillar.AddNode("Yellow");
            }));

            prompt.AddChoice(new Vehicle("Krauss-Maffei Leopard 2", rootKraussMaffei =>
            {
                // Add some nodes
                var fooKraussMaffei = rootKraussMaffei.AddNode("[yellow]Engine[/]");
                fooKraussMaffei.AddNode("12-Zylinder-Dieselengine MTU");
                fooKraussMaffei.AddNode("CO2 emissions don't care");

                var barKraussMaffei = rootKraussMaffei.AddNode("[yellow]Military equipment[/]");
                barKraussMaffei.AddNode("Laser sword");
                barKraussMaffei.AddNode("Full rotating tower");
            }));
            
            var rentalVehicle = AnsiConsole.Prompt(prompt);
            AnsiConsole.MarkupLine($"You selected [green]{rentalVehicle.Title}[/]");

        }
    }
}
```

(The `Action<Tree>` was only chosen to keep similarity to the code above, I would probably use the tree internally only and have some properties to be set on the `Vehicle`.)

**Reply by DiagProf** [reply]
Thanks for the detailed answer.  
That helps me alot.
